### PR TITLE
Define _REENTRANT for SUNOS

### PR DESCRIPTION
**Pull request by AustinWise**
This [matches mono](https://github.com/dotnet/runtime/blob/main/src/mono/CMakeLists.txt#L295). This is required to make errno a thread-local variable. This fixes problems where calls into `libSystem.Native` return failure but the the `errno` is 0. This was causing all kinds of odd asserts in System.IO.Tests and other libraries tests.

See the man page for  [errno(3c)](https://illumos.org/man/3C/errno):

> In the case of multithreaded applications, the `-mt` option must be specified on the command line at compilation time (see threads(7)). When the `-mt` option is specified, `errno` becomes a macro that enables each thread to have its own `errno`.

Since `-mt` is presumably specific to an old Sun compiler, the [threads(7)](https://illumos.org/man/7/threads) man page says:

> Users of other compilers such as gcc and clang should manually set `-D_REENTRANT` on the compilation line. There are no other libraries flags necessary.

Contributes to #34944

### CODE REVIEW

**Comment by am11** [comment]
I hope you won't take it the wrong way, and I am happy this port is moving forward. I think it'd be better to port something sizeable, such as one of the remaining runtime library needed to complete shared framework, rather than many teeny tiny patches. Otherwise, we would be burdening the maintainers which will slow down the port process / interest.

Thanks for pushing it forward.

**Review by am11 — APPROVED** [review]
LGTM, it is aligning with mono: https://github.com/dotnet/runtime/blob/9882188fbe16adee3538b65a3926465dfd6052c1/src/mono/CMakeLists.txt#L295

File: eng/native/configurecompiler.cmake [hunk 1]
```diff
@@ -298,7 +298,7 @@ elseif(CLR_CMAKE_HOST_SUNOS)
   add_compile_options($<$<COMPILE_LANGUAGE:ASM>:-Wa,--noexecstack>)
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fstack-protector")
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fstack-protector")
-  add_definitions(-D__EXTENSIONS__ -D_XPG4_2 -D_POSIX_PTHREAD_SEMANTICS)
+  add_definitions(-D__EXTENSIONS__ -D_XPG4_2 -D_POSIX_PTHREAD_SEMANTICS  -D_REENTRANT)
```
Comment (am11) [inline eng/native/configurecompiler.cmake:301]:
```suggestion
  add_definitions(-D__EXTENSIONS__ -D_XPG4_2 -D_POSIX_PTHREAD_SEMANTICS -D_REENTRANT)
```

**Review by gwr — APPROVED** [review]

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Pull request merged**